clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -251,42 +253,45 @@ public int partitionLeaderEpoch() {
         return buffer.getInt(PARTITION_LEADER_EPOCH_OFFSET);
     }
 
-    private CloseableIterator<Record> compressedIterator(BufferSupplier bufferSupplier) {
+    private CloseableIterator<Record> compressedIterator(BufferSupplier bufferSupplier, boolean skipKeyValue) {
         final ByteBuffer buffer = this.buffer.duplicate();
         buffer.position(RECORDS_OFFSET);
         final DataInputStream inputStream = new DataInputStream(compressionType().wrapForInput(buffer, magic(),
-                bufferSupplier));
-
-        return new RecordIterator() {
-            @Override
-            protected Record readNext(long baseOffset, long firstTimestamp, int baseSequence, Long logAppendTime) {
-                try {
-                    return DefaultRecord.readFrom(inputStream, baseOffset, firstTimestamp, baseSequence, logAppendTime);
-                } catch (EOFException e) {
-                    throw new InvalidRecordException("Incorrect declared batch size, premature EOF reached");
-                } catch (IOException e) {
-                    throw new KafkaException("Failed to decompress record stream", e);
-                }
-            }
-
-            @Override
-            protected boolean ensureNoneRemaining() {
-                try {
-                    return inputStream.read() == -1;
-                } catch (IOException e) {
-                    throw new KafkaException("Error checking for remaining bytes after reading batch", e);
+            bufferSupplier));
+
+        if (skipKeyValue) {
+            // this buffer is used to skip length delimited fields like key, value, headers
+            byte[] array = new byte[MAX_SKIP_BUFFER_SIZE];
+            ByteBuffer skipBuffer = ByteBuffer.wrap(array);
+            // set its limit to 0 to indicate no bytes readable yet
+            skipBuffer.limit(0);
+
+            return new StreamRecordIterator(inputStream) {
+                @Override
+                protected Record readNext(long baseOffset, long firstTimestamp, int baseSequence, Long logAppendTime) {
```

Comment:
nit: The only difference between this branch and the other one is the `readPartiallyFrom` line. I think we could move the exception handling to the abstract class. You can add a method like this:
```java
abstract Record doReadRecord(
  long baseOffset, 
  long firstTimestamp, 
  int baseSequence, 
  Long logAppendTime) throws IOException;
```

Reply:
Ack.